components/l10n/common/locale_util.cc

```diff
@@ -5,78 +5,114 @@
 
 #include "brave/components/l10n/common/locale_util.h"
 
-#include <algorithm>
-#include <vector>
-
-#include "base/strings/string_split.h"
-#include "base/strings/string_util.h"
-#include "base/strings/utf_string_conversions.h"
-#include "ui/base/resource/resource_bundle.h"
+#include "absl/types/optional.h"
+#include "base/check.h"
+#include "base/no_destructor.h"
+#include "brave/components/l10n/common/locale_constants.h"
+#include "servloc.h"
+#include "third_party/icu/source/common/unicode/locid.h"
+#include "ui/base/l10n/l10n_util.h"
 
 namespace brave_l10n {
 
 namespace {
 
-const char kDefaultLanguageCode[] = "en";
-const char kDefaultCountryCode[] = "US";
+absl::optional<icu::Locale>& ICULocaleForTesting() {
+  static base::NoDestructor<absl::optional<icu::Locale>> icu_locale;
+  return *icu_locale;
+}
 
-}  // namespace
+absl::optional<icu::Locale>& LastICULocaleForTesting() {
+  static base::NoDestructor<absl::optional<icu::Locale>> last_icu_locale;
+  return *last_icu_locale;
+}
 
-std::string GetLanguageCode(
-    const std::string& locale) {
-  const std::vector<std::string> locale_components = base::SplitString(locale,
-      ".", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+inline icu::Locale GetNormalizedICULocale(const std::string& locale) {
+  return {l10n_util::NormalizeLocale(locale).c_str()};
+}
 
-  if (locale_components.empty()) {
-    return kDefaultLanguageCode;
+inline std::string GetLocale(const icu::Locale& icu_locale) {
+  std::string locale{icu_locale.getBaseName()};
+  if (locale.empty()) {
+    return kDefaultLocale;
   }
 
-  std::string normalized_locale = locale_components.front();
-  std::replace(normalized_locale.begin(), normalized_locale.end(), '-', '_');
-
-  const std::vector<std::string> components = base::SplitString(
-      normalized_locale, "_", base::KEEP_WHITESPACE, base::SPLIT_WANT_ALL);
+  return locale;
+}
 
-  if (components.empty()) {
+inline std::string GetLanguageCode(const icu::Locale& icu_locale) {
+  std::string language_code{icu_locale.getLanguage()};
+  if (language_code.empty()) {
     return kDefaultLanguageCode;
```

Comment:
same as above, I'm concerned here that we're using hard-coded defaults here that aren't from upstream l10n

Reply:
Discussed via DM. LocaleHelper has always returned en_US for the default. Would be good to follow up in another PR to investigate/use GetApplicationLocale.